prometheus/promauto/auto.go

```diff
@@ -132,92 +164,211 @@ import "github.com/prometheus/client_golang/prometheus"
 func NewGaugeFunc(opts prometheus.GaugeOpts, function func() float64) prometheus.GaugeFunc {
-	g := prometheus.NewGaugeFunc(opts, function)
-	prometheus.MustRegister(g)
-	return g
+	return With(prometheus.DefaultRegisterer).NewGaugeFunc(opts, function)
 }
 
 // NewSummary works like the function of the same name in the prometheus package
 // but it automatically registers the Summary with the
 // prometheus.DefaultRegisterer. If the registration fails, NewSummary panics.
 func NewSummary(opts prometheus.SummaryOpts) prometheus.Summary {
-	s := prometheus.NewSummary(opts)
-	prometheus.MustRegister(s)
-	return s
+	return With(prometheus.DefaultRegisterer).NewSummary(opts)
 }
 
 // NewSummaryVec works like the function of the same name in the prometheus
 // package but it automatically registers the SummaryVec with the
 // prometheus.DefaultRegisterer. If the registration fails, NewSummaryVec
 // panics.
 func NewSummaryVec(opts prometheus.SummaryOpts, labelNames []string) *prometheus.SummaryVec {
-	s := prometheus.NewSummaryVec(opts, labelNames)
-	prometheus.MustRegister(s)
-	return s
+	return With(prometheus.DefaultRegisterer).NewSummaryVec(opts, labelNames)
 }
 
 // NewHistogram works like the function of the same name in the prometheus
 // package but it automatically registers the Histogram with the
 // prometheus.DefaultRegisterer. If the registration fails, NewHistogram panics.
 func NewHistogram(opts prometheus.HistogramOpts) prometheus.Histogram {
-	h := prometheus.NewHistogram(opts)
-	prometheus.MustRegister(h)
-	return h
+	return With(prometheus.DefaultRegisterer).NewHistogram(opts)
 }
 
 // NewHistogramVec works like the function of the same name in the prometheus
 // package but it automatically registers the HistogramVec with the
 // prometheus.DefaultRegisterer. If the registration fails, NewHistogramVec
 // panics.
 func NewHistogramVec(opts prometheus.HistogramOpts, labelNames []string) *prometheus.HistogramVec {
+	return With(prometheus.DefaultRegisterer).NewHistogramVec(opts, labelNames)
+}
+
+// NewUntypedFunc works like the function of the same name in the prometheus
+// package but it automatically registers the UntypedFunc with the
+// prometheus.DefaultRegisterer. If the registration fails, NewUntypedFunc
+// panics.
+func NewUntypedFunc(opts prometheus.UntypedOpts, function func() float64) prometheus.UntypedFunc {
+	return With(prometheus.DefaultRegisterer).NewUntypedFunc(opts, function)
+}
+
+// Factory provides factory methods to create Collectors that are automatically
+// registered with a Registerer. Create a Factory with the With function,
+// providing a Registerer to auto-register created Collectors with. The zero
+// value of a Factory creates Collectors that are not registered with any
+// Registerer. All methods of the Factory panic if the registration fails.
+type Factory struct {
+	r prometheus.Registerer
+}
+
+// With creates a Factory using the provided Registerer for registration of the
```

Comment:
What if registerer is nil? (We often do that for tests)

Reply:
Oh actually it is just not registered... Awesome! :heart:

Reply:
That's also documented in the `Factory` doc comment.